### ISSUE TITLE
Continuous Integration

### ISSUE THREAD

**Issue by RAnders00** [jfoenixadmin/JFoenix#58]
Currently, the JFoenix binaries are compiled and updated once in a while. I think this project would benefit from continuous integration, which means that a service builds and tests the project on every push to the GitHub repository. That way, you would have fresh binaries for every repository version (which is useful, since users are then able to get bleeding-edge builds to reproduce issues without having to build on their own). Also, if configured correctly, pull requests can also be tested whether their unit tests pass. I would be happy to help with setting up the CI, in case help is needed. I would recommend [CircleCI](https://www.circleci.com/), since I can provide help with it and it also allows to save build artifacts for every build in history (which is not the case for all services).

Since this project is open source, we can benefit from free plans at a lot of services.

**Comment by jfoenixadmin** [comment]
it's good to benefit from continuous integration free plans, your help would be appreciated with CircleCI. Thanks

**Comment by RAnders00** [comment]
I will prepare a pull request.

**Comment by jfoenixadmin** [comment]
i tested it, it's working. is it possible to automate the maven publish using CircleCI? also publish to jfoenix site?

Thanks

**Comment by RAnders00** [comment]
I will get around to doing that. I just need to know: are you directly shipping to Maven Central or via JCenter?

**Comment by jfoenixadmin** [comment]
we are shipping directly to Maven Central, and the files are uploaded manually to the website.

**Comment by aalmiray** [comment]
FWIW the gradle build already defines basic setup for publishing artifacts to Bintray's JCenter (see #22). Artifacts can be synced automatically from Bintray to Maven Central.

Besides Circle-CI, you could also try Appveyor (Windows build, in case the project requires Windows specific tests) and Travis-CI (perhaps the easiest one to setup). I wouldn't recommend pushing to a binary repository after every successful unless your intention is to generate a lot of snapshot releases.

May I suggest to have a look at http://semver.org/ for getting ideas on how versions and releases may be handled? Semver rules are used by many projects out there, and have been recognized by many as a good set of rules to follow.

**Comment by RAnders00** [comment]
I appreciate you listing all these alternatives, but this project really does not need a CI except for making binaries for every version. I already looked into Travis-CI and they don't offer to save artifacts for every build. CircleCI offers saving a list of binaries for every build.

Also, I agree on the semantic versioning point. This project could really be versioned (currently, everything is in "snapshot forever" state and everything gets the version 0.0.0-SNAPSHOT).

I'm not quite into the different online maven repositories, but would it be an appropriate solution to push every build to a snapshot repository and every full release (that could be distinguished once we have versioning) could be uploaded to JCenter with Maven Central synchronization?

**Comment by aalmiray** [comment]
Right. As far as I can tell you can invoke any commands after a successful build in Travis-CI, for example an automatic push to Maven Central's Snapshot repository or Bintray's (or both). In this way you don't have to expect the CI environment to keep the built artifacts around.

I understand the reasoning behind keeping a "snapshot forever" version, as it keeps things simple for the development team, but can lead to confusion for your library's consumers. As I recall the JAR manifest includes the build time/date however it could also include the SCM commit hash; in this way it would be easier to distinguish which features are available in to snapshot JARs. I'll send a PR with this behavior shortly.

In summary, the build file can be configured in such a way that snapshot releases are automatically published when built in CI. A versioned released can be releases using the same mechanism but I'd recommend doing this manually (so that you can update the version number according to your liking).

**Comment by RAnders00** [comment]
CircleCI can also do that. in `circle.yml`:

```yaml
deployment:
  mavenCentral:
    branch: master
    commands:
     - echo first command
     - echo second command
     - ./gradlew deployToMavenCentral # or something along those lines
```

That will only run if the build was successful.

[the docs are here](https://circleci.com/docs/configuration#deployment)
